File: src/main/java/com/github/puregero/warpplugin/SetWarpCommand.java
package com.github.puregero.warpplugin;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.Location;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

public class SetWarpCommand implements CommandExecutor {
    private final WarpPluginPlugin plugin;

    public SetWarpCommand(WarpPluginPlugin plugin) {
        this.plugin = plugin;
        Objects.requireNonNull(plugin.getCommand("setwarp")).setExecutor(this);
    }

    @Override
    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command, @NotNull String label, @NotNull String[] args) {
        if (args.length == 0) {
            sender.sendMessage(Component.text("Usage: /%s <warp>".formatted(label)).color(NamedTextColor.RED));
            return false;
        }

        String warpName = args[0];

        if (!(sender instanceof Player player)) {
            sender.sendMessage(Component.text("Only players can execute this command!").color(NamedTextColor.RED));
            return false;
        }

        Location location = player.getLocation();

        synchronized (this.plugin) {
            ConfigurationSection warpSection = this.plugin.getConfig().createSection("warps." + warpName);
            warpSection.set("world", location.getWorld().getName());
            warpSection.set("x", location.getBlockX() + 0.5);
            warpSection.set("y", location.getY());
            warpSection.set("z", location.getBlockZ() + 0.5);
            warpSection.set("yaw", (int) ((location.getYaw() + 45) / 90) * 90); // round to nearest 90 degrees
            warpSection.set("pitch", 0);
        }

        this.plugin.scheduleConfigSave();

        player.sendMessage(Component.text("Warp set: " + warpName).color(NamedTextColor.GREEN));

        return true;
    }
}

File: src/main/java/com/jjkay03/foliawarp/commands/SetWarpCommand.kt
package com.jjkay03.foliawarp.commands

import com.jjkay03.foliawarp.FoliaWarp
import com.jjkay03.foliawarp.Utils
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

class SetWarpCommand : CommandExecutor {

    init {
        FoliaWarp.INSTANCE.getCommand("setwarp")?.setExecutor(this)
    }

    // COMMAND
    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<out String>): Boolean {
        // End if invalid arg
        if (args.isEmpty()) { sender.sendMessage("§cUsage: /$label <warp>"); return false }

        val warpName = args[0]

        // End if not player
        if (sender !is Player) { sender.sendMessage("§cOnly players can execute this command!"); return false }

        // Create warp
        val location = sender.location
        synchronized(FoliaWarp.INSTANCE) {
            val warpSection = FoliaWarp.INSTANCE.config.createSection("warps.$warpName")
            warpSection.set("world", location.world.name)
            warpSection.set("x", location.blockX + 0.5)
            warpSection.set("y", location.y)
            warpSection.set("z", location.blockZ + 0.5)
            warpSection.set("yaw", ((location.yaw + 45) / 90).toInt() * 90) // round to nearest 90 degrees
            warpSection.set("pitch", 0)
        }

        // Save warp in config
        Utils.scheduleConfigSave()

        // Feedback
        sender.sendMessage("§3\uD83C\uDF00 §bSet warp §3$warpName")

        return true
    }

}
